### ISSUE TITLE
Enhancement: add encoding/decoding half precision floating point functions to public API

### ISSUE THREAD

**Issue by phirsov** [intel/tinycbor#149]
It's uncommon to operate half precision floating point data on application level, while such format is widely used on transport level to minimize traffic. In such case one can need API for encoding/decoding half-precision floating point data AS more common single or double. Something like this:
```c
CborError cbor_encode_float_as_half_float(CborEncoder *encoder, float value);
CborError cbor_encode_double_as_half_float(CborEncoder *encoder, double value);
CborError cbor_value_get_half_float_as_float(const CborValue *value, float *result);
CborError cbor_value_get_half_float_as_double(const CborValue *value, double *result);
// etc.
```
There is some related stuff in the private area, and I think, I can make it public.

**Comment by phirsov** [comment]
[Partially implemented](https://github.com/phirsov/tinycbor/commit/62b22ee4781a815c1b991a8215aa32587b6acebe). Is this OK. or some amendment is needed?

**Comment by thiagomacieira** [comment]
Yes, it looks very good. I have one overall comment and two very minor. Overall, I think the `float` API has more value than `double`. Any system capable of double-precision either already has a library function to convert single to double or the FPU implements it. But there are systems without `double` support, or at least more costly than `float`. So let's explore changing the internal function and the API to return `float` and receive `float`.

Thanks for the tests, they look nice.

The two minor comments:

1. Name the two files `cborparser_float.c` and `cborenccoder_float.c` please
2. Add the two files to tst_cpp.

**Comment by phirsov** [comment]
[Look at the improvement made so far](https://github.com/intel/tinycbor/compare/master...phirsov:get-half-as). But I'm in doubt now: I have used ldexpf function, which is not accessible in plain old C89 and some antique compilers.

**Comment by thiagomacieira** [comment]
Let's do bit manipulation directly. I wonder if there's a `#define` that helps us know that FP is IEEE 754. C++ has that.

**Comment by phirsov** [comment]
```cpp
static inline float decode_halff(unsigned short half)
{
    int exp = (half >> 10) & 0x1f;
    int mant = half & 0x3ff;
    float mantf, expf, val;
    if (exp == 0) {
        mantf = mant;
        expf = 1.0f / (1 << 24);
        val = mantf * expf;
    } else if (exp != 31) {
        mantf = mant + 1024.0f;
        expf = exp >= 25 ? 1 << (exp - 25) : 1.0f / (1 << (25 - exp));
        val = mantf * expf;
    } else {
        val = mant == 0 ? INFINITY : NAN;
    }
    return half & 0x8000 ? -val : val;
}
```

**Comment by thiagomacieira** [comment]
If that produces proper results, that's fine. I was thinking of converting the `uint16_t`  to `uint32_t` first, then `memcpy` from that to a `float`, instead of doing FP calculations. If we're going to do math, I'd rather use `ldexpf`, which according to the standard is part of C99.

**Comment by phirsov** [comment]
So, there are 4 ways with their pro-s and contra-s:

1. use ldexp: standard-compliant, works on most platforms, but can be expensive;
1. use ldexpf: needs the C99 compiler, can be less expensive (but still expensive);
1. use naive math: works without std math lib (yes, sometimes linking without -lm option saves a lot of space), but error-prone and still expensive;
1. direct bit manipulations: re-shifting mantissa, re-biasing exponent... Error-prone again, cheap for platforms without hardware floating-point math. But what's about platforms with non-IEEE-754 math? Well. we can end with dropping float and double from API at all:
```c
#ifndef CBOR_NO_PLATFORM_IEEE754_FLOATING_POINT
CBOR_INLINE_API CborError cbor_value_get_float(CborEncoder *encoder, float *value);
CBOR_INLINE_API CborError cbor_encode_float(CborEncoder *encoder, float value);
// ... for half and double
#else
// Caveat lector: IEEE754 floating point representation is assumed
CBOR_INLINE_API CborError cbor_value_get_float(CborEncoder *encoder, uint32_t *value);
CBOR_INLINE_API CborError cbor_encode_float(CborEncoder *encoder, uint32_t value);
// ... for half and double
#endif
```

What is the preferable way?

**Comment by thiagomacieira** [comment]
CBOR is defined as carrying IEEE754 binary64, binary32 and binary16 content. So I like your idea.

I don't think we need to provide the `uintXX_t` version of the APIs for non-IEEE754 systems, though. Let people use `cbor_encode_floating_point` and `cbor_value_get_floating_point` (the latter does not exist), which operate on `void *`.

**Comment by phirsov** [comment]
OK, but I think, this is the task for separate feature request. As for now, let's roll back to plain old doubles (ldexp, not ldexpf) both for cbor_value_get_half_float_as_float and cbor_value_get_half_float_as_double.

**Comment by phirsov** [comment]
Well, cbor_value_get_half_float_as_float needs a pointer to float, so user should do extra effort. Hope, nobody could try something like:
```c
double x;
/* Oops... I know a guy, who can try it without any doubt */
cbor_value_get_half_float_as_float(cbor_value, (float*)&x);
```
But in name of API brevity `..._as_double` should be eliminated.

**Comment by thiagomacieira** [comment]
If someone wants to shoot themselves in the foot, who are we to say no? :-)

**Comment by phirsov** [comment]
[Implemented](https://github.com/intel/tinycbor/compare/master...phirsov:get-half-as). I'll squash the branch into the single commit before pull request.

**Comment by x448** [comment]
I recently implemented IEEE 754 binary16 conversion functions for use by a cbor library. It tests all 4+ billion conversions using IEEE 754 default rounding ("Round-to-Nearest RoundTiesToEven") which is the most accurate and statistically unbiased estimate of the true result. 

* https://github.com/cbor-go/float16 (written by me)
* https://github.com/fxamacker/cbor (written by Faye Amacker)

There are pitfalls regarding NaN conversions because of silent sNaN->qNaN conversion when casting floats (both C and Go).  So I provide a separate function for converting NaN that can preserve signal bit in case protocols using CBOR need it.

To document what it means to "preserve value" and how to handle subnormals, I wrote the following definition for use by fxamacker/cbor:

> For generic encoders and floating-point numbers, "preserves its value" means being able to round-trip back to the original floating-point value with all underlying bits unchanged. For example, encode float32 values as float16 only if a round-trip of float32->float16->float32 produces the original float32 with all 32 underlying bits unchanged. This includes 2046 subnormal float32 values that can successfully round-trip in the manner described. Rounding for float64->float32 and float32->float16 use IEEE 754 default rounding ("Round-to-Nearest RoundTiesToEven") which is the most accurate and statistically unbiased estimate of the true result. Conversions from smaller sizes to larger sizes are lossless and don't require rounding. Conversions from float64->float16 must produce the same result as float64->float32->float16.

I hope you find this useful.  I'm finding it helpful to view CBOR as a binary format construction kit rather than a specific binary format because a lot of decisions are left to protocol designers.

**Comment by thiagomacieira** [comment]
Hello @x448 
we have the code to encode and decode half floats. I just didn't want to add an API that uses that.

**Comment by x448** [comment]
@thiagomacieira Oops!   Deleted.  Thanks for letting me know.
